algebra-core/src/curves/mod.rs

```diff
@@ -308,3 +310,218 @@ where
+                    &mut tmp,
+                    &mut a_2.to_vec()[..],
+                    (0..batch_size).map(|x| (x, x)).collect(),
+                );
+            }
+
+            for (table, p) in tables.iter_mut().zip(&tmp) {
+                table.push(p.clone());
+            }
+        }
+        tables
+    }
+
+    // This function mutates the scalars in place
+    // We can make this more generic in the future to use other than u16.
+    fn batch_wnaf_opcode_recoding<BigInt: BigInteger + AsRef<[u64]>>(
+        scalars: &mut [BigInt],
+        w: usize,
+    ) -> Vec<Vec<Option<i16>>> {
+        assert!(w > 0);
+        let batch_size = scalars.len();
+        let window_size: i16 = 1 << (w + 1);
+        let half_window_size: i16 = 1 << w;
+
+        let mut op_code_vectorised =
+            Vec::<Vec<Option<i16>>>::with_capacity(scalars[0].as_ref().len() * 64);
+
+        let mut all_none = false;
+        while !all_none {
+            let mut opcode_row = Vec::with_capacity(batch_size);
+
+            for s in scalars.iter_mut() {
+                if s.is_zero() {
+                    opcode_row.push(None);
+                } else {
+                    let op = if s.is_odd() {
+                        let mut z: i16 = (s.as_ref()[0] % (1 << (w + 1))) as i16;
+
+                        if z < half_window_size {
+                            s.sub_noborrow(&BigInt::from(z as u64));
+                        } else {
+                            z = z - window_size;
+                            s.add_nocarry(&BigInt::from((-z) as u64));
+                        }
+                        z
+                    } else {
+                        0
+                    };
+                    opcode_row.push(Some(op));
+                    s.div2();
+                }
+            }
+
+            all_none = opcode_row.iter().all(|x| x.is_none());
+            if !all_none {
+                op_code_vectorised.push(opcode_row);
+            }
+        }
+        op_code_vectorised
+    }
+
+    // This function consumes the second op as it mutates it in place
+    // to prevent memory allocation
+    fn batch_double_in_place(bases: &mut [Self], index: Vec<usize>);
+
+    fn batch_add_in_place(bases: &mut [Self], other: &mut [Self], index: Vec<(usize, usize)>);
+
+    fn batch_scalar_mul_in_place<BigInt: BigInteger>(
+        mut bases: &mut [Self],
+        scalars: &mut [BigInt],
+        w: usize,
+    ) {
+        let opcode_vectorised = Self::batch_wnaf_opcode_recoding::<BigInt>(scalars, w);
+        let tables = Self::batch_wnaf_tables(bases, w);
+
+        // Set all points to 0;
+        let zero = Self::zero();
+        for p in bases.iter_mut() {
+            *p = zero;
+        }
+
+        for opcode_row in opcode_vectorised.iter().rev() {
+            let index_double = opcode_row
+                .iter()
+                .enumerate()
+                .filter(|x| x.1.is_some())
+                .map(|x| x.0)
```

Comment:
Maybe replace this with [`filter_map`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.filter_map)?